vinumc: improve text handling

JeanJPNM opened the pull request:
Closes #1.
Fixes #38.
Closes #41.
Closes #5.
Fixes #70.

- [x] update grammar to take whitespace into account
- [x] perform blank reduction in eval
- [x] literal delimiter
- [x] character escaping
- [x] write tests for the literal delimiter
- [x] write tests for character escaping

Grillo-0 commented on subprojects/vinumc/dry_flex.l:
```diff
@@ -18,11 +19,20 @@
+
+	return WORD;
+}
 
```
@maju-degrandi can you review this lex/bison part for me? I think you know more of this than me XD

Grillo-0 commented on subprojects/vinumc/str.c:
```diff
@@ -9,3 +11,47 @@ void put_str(struct str *to, char *from) {
+
+	bool is_previous_blank = to->len == 0 || is_blank_char(VEC_AT(to, to->len - 1));
+
+	// TODO: should probably use VEC_RESERVE
```
Yeah, just use it

JeanJPNM replied:
`VEC_RESERVE` has a problem, since it allocates exactly the amount of elements specified, calling it repeatedly (which _will_ happen a lot in `eval`) will have worse performance than just using `VEC_PUT` directly, since `VEC_PUT` expands the vector capacity exponentially rather than linearly.

Grillo-0 replied:
I don't think this is correct.

Imagine you have an `out` with `out->capacity == 16` and `out->len == 16` and you're going to insert `48` new items.

Using `VEC_RESERVE` you're going to do one `realloc` (`16 -> 64`).

Without using `VEC_RESERVE` you're going to need two `realloc`'s (`16->32->64`)

Please correct me if I'm wrong :)

Grillo-0 replied:
Oh, after thinking more about it, I understood where you came from.

The `out` is going to be the same between calls, and using `VEC_RESERVE` will garantee an allocation on every call. Yeah, you're right, for now let's not use `VEC_RESERVE`

Grillo-0 commented on subprojects/vinumc/dry_flex.l:
```diff
@@ -15,58 +21,109 @@
+}
+
+<CONTENT>([^\[\]\\]) {
+	// covers case where the file ends abruptly mid word
```
Create a test for that

JeanJPNM replied:
I'm not sure about how to do that, because this would essentially be a test in which the compiler receives a malformed input (which, as of now, would not be added to the ast and thus, not evaluated). I also can't remove this pattern because it would make the flex scanner jam when it runs into that scenario.

Grillo-0 commented on subprojects/vinumc/eval.h:
```diff
@@ -14,6 +14,12 @@ enum entry_type {
+	LAST_CHILD = 1 << 1,
+	REDUCE_BLANKS = 1 << 2,
+};
+
```
Make this internal to the C file

Grillo-0 commented on subprojects/vinumc/str.c:
```diff
@@ -12,16 +12,28 @@ void put_str(struct str *to, char *from) {
 
-void put_start_blank_reduced_str(struct str *to, char *from) {
-	size_t len = strlen(from);
+void put_blank_reduced_str(struct str *to, char *from, bool skip_left, bool skip_right) {
```
I think this function could use some documentation. As we don't yet have a formalized way of documentation, just add a comment above explaining what this function does

Grillo-0 commented on subprojects/vinumc/str.c:
```diff
@@ -12,16 +12,28 @@ void put_str(struct str *to, char *from) {
 		VEC_POP(to);
 	}
 
+	size_t start = 0;
+	size_t end = strlen(from);
+	if (skip_right) {
+		while (end > 0 && is_blank_char(from[end - 1])) {
+			end--;
+		}
+	}
+	if (skip_left) {
+		while (start < end && is_blank_char(from[start])) {
+			start++;
+		}
+	}
```
couldn't this been added when you first created the function?

JeanJPNM replied:
I don't understand what you mean. Could you elaborate?

Grillo-0 commented on subprojects/vinumc/tests/library_test.c:
```diff
@@ -6,7 +6,7 @@ void test_call(struct vunit_test_ctx *ctx) {
 			    "subprojects/vinumc/tests/libtestlib.so", NULL);
 
-	VUNIT_ASSERT_STREQ(ctx, out, "test\n");
+	VUNIT_ASSERT_STREQ(ctx, out, "test");
```
Squash this commit into the previous one, the test probably fail on the previous commit

Grillo-0 commented on subprojects/vinumc/dry_flex.l:
```diff
@@ -23,8 +24,25 @@
+		text_buffer = (struct str){};                                                      \
+		yylval = ast_add_node(&ctx.ast, node);                                             \
+		return TEXT;                                                                       \
+	} while (0)
```
just use a function for that

Grillo-0 commented on subprojects/vinumc/dry_flex.l:
```diff
@@ -14,30 +22,117 @@
+// used to handle character escaping inside the regular text blocks
+// preventing having two consecutive TEXT nodes
+// that could be just one node
+struct str text_buffer = {};
```
```suggestion
static struct str text_buffer = {};
```

Grillo-0 reviewed (changes requested):
Update the version number

JeanJPNM committed:
vinumc: tokenize by text instead of word

This commit removes the WORD token, and puts all
text content (including whitespace) into the TEXT
token.
This also means that eval is now responsible for
performing blank reduction.

JeanJPNM committed:
vinumc/dry_flex: implement character escaping

JeanJPNM committed:
vinumc: implement literal delimiter

JeanJPNM commented on subprojects/vinumc/eval.c:
```diff
@@ -242,35 +253,33 @@ DO_CALLS_FUNC_SIGNATURE(do_calls_call) {
 
 		// put the extern function call return on the out str
 		struct return_value call_return = symbol_info->as.func(tmp_out.base);
-		put_str(out, call_return.ptr);
+
+		if ((flags & REDUCE_BLANKS) != 0) {
+			put_blank_reduced_str(out, call_return.ptr, true, true);
+		} else {
+			put_str(out, call_return.ptr);
+		}
```
The vinum specification doesn't say if the output of external functions should be trimmed, so I decided make them match the behavior of regular functions.

Grillo-0 commented on subprojects/vinumc/str.h:
```diff
@@ -2,9 +2,17 @@
 #define __STR_H__
 
 #include "vec.h"
+#include <stdbool.h>
```
```suggestion
#include <stdbool.h>

#include "vec.h"
```

Grillo-0 commented on subprojects/vinumc/str.c:
```diff
@@ -1,6 +1,8 @@
 #include "str.h"
 #include <string.h>
 
+static bool is_blank_char(char c);
+
```
Just place the function here

Grillo-0 commented on subprojects/vinumc/str.h:
```diff
@@ -2,9 +2,17 @@
 
 void put_str(struct str *to, char *from);
 
+// Puts the contents of `from` into `to`, performing blank reduction,
+// that is, collapsing consecutive blank characters into a single space.
+//
+// If `trim_left` is true, leading blanks are removed.
+// If `trim_right` is true, trailing blanks are removed.
```
place this on the .c

Grillo-0 reviewed (changes requested):
bump the version number

JeanJPNM commented on subprojects/vinumc/str.c:
```diff
@@ -9,3 +14,46 @@ void put_str(struct str *to, char *from) {
+		char c = from[i];
+
+		if (!is_blank_char(c)) {
+			VEC_PUT(to, c);
```
Is using `VEC_PUT` here despite calling `VEC_RESERVE` ok? I'm not sure if it is preferrable to write `to->base[to->len++] = c;` instead.

Grillo-0 replied:
It's ok for me

arthurvergacas reviewed (approved):
LGTM
